packages/lisk-transactions/src/base_transaction.ts

```diff
@@ -0,0 +1,497 @@
+				) {
+					return [
+						...errorArray,
+						new TransactionError('Missing signatures', this.id, '.signatures'),
+					];
+				}
+
+				// Check senderId
+				if (
+					key === 'address' &&
+					typeof value === 'string' &&
+					value.toUpperCase() !== this.senderId.toUpperCase()
+				) {
+					return [
+						...errorArray,
+						new TransactionError(
+							'Invalid sender address',
+							this.id,
+							'.senderId',
+						),
+					];
+				}
+
+				return errorArray;
+			},
+			[],
+		);
+
+		// Check missing secondPublicKey on account
+		const missingSecondPublicKeyError =
+			this.signSignature && !sender.secondPublicKey
+				? [
+						new TransactionError(
+							'Sender does not have a secondPublicKey',
+							this.id,
+						),
+				  ]
+				: [];
+
+		// Signature verifications
+
+		const transactionHash = cryptography.hash(this.getBasicBytes());
+
+		// Verify signSignature
+		const signSignatureError =
+			sender.secondPublicKey &&
+			!cryptography.verifyData(
+				transactionHash,
+				this.signSignature as string,
+				sender.secondPublicKey,
+			)
+				? [
+						new TransactionError(
+							'Failed to verify second signature',
+							this.id,
+							'.signSignature',
+						),
+				  ]
+				: [];
+
+		// Verify multisignatures
+		const transaction = this.toJSON();
+		const unvalidatedSignatures = validateMultisignatures(
+			sender,
+			transaction,
+			transactionHash,
+		);
+
+		const multisignatureVerificationErrors =
+			unvalidatedSignatures.length > 0
+				? unvalidatedSignatures.map(
+						signature =>
+							new TransactionError(
+								`Failed to verify multisignature: ${signature}`,
+								transaction.id,
+								'.signatures',
+							),
+				  )
+				: [];
+
+		const verifyErrors: ReadonlyArray<TransactionError> = [
+			...balanceError,
+			...transactionErrors,
+			...missingSecondPublicKeyError,
+			...signSignatureError,
+			...multisignatureVerificationErrors,
+		];
+
+		return {
+			id: this.id,
+			status: verifyErrors.length === 0 ? Status.OK : Status.FAIL,
+			errors: verifyErrors,
+		};
+	}
+
+	public abstract verifyAgainstOtherTransactions(
+		transactions: ReadonlyArray<TransactionJSON>,
+	): TransactionResponse;
+
+	public apply(sender: Account): TransactionResponse {
+		const updatedBalance = new BigNum(sender.balance).sub(this.fee);
+		const updatedAccount = { ...sender, balance: updatedBalance.toString() };
+
+		return {
+			id: this.id,
+			status: Status.OK,
+			state: [updatedAccount],
+			errors: [],
+		};
+	}
+
+	public undo(sender: Account): TransactionResponse {
+		const updatedBalance = new BigNum(sender.balance).add(this.fee);
+		const updatedAccount = { ...sender, balance: updatedBalance.toString() };
+
+		return {
+			id: this.id,
+			status: Status.OK,
+			state: [updatedAccount],
+			errors: [],
+		};
+	}
+
+	public isExpired(): boolean {
+		// tslint:disable-next-line no-magic-numbers
+		const timeNow = Math.floor(Date.now() / 1000);
```

Comment:
is it important to divide by 1000?
also, @SargeKhan wanted to make it like
`isExpired(now: Date = Date.now()): boolean`?

Reply:
Yes, I think the function should accept the date parameter.

Reply:
@shuse2 `/  1000` follows the logic in core `transaction_pool:isExpired`